{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyvcf2 import VCF\n",
    "import pandas as pd\n",
    "import time\n",
    "import multiprocessing\n",
    "from datetime import date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_columns(df):\n",
    "    data_1 = df.copy()\n",
    "    data_1['REF'] = '.'\n",
    "    data_1['ALT'] = '.'\n",
    "    data_1['GENOTYPE'] = '.'\n",
    "    \n",
    "    return data_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunk(L,nchunks):\n",
    "    L2 = list()\n",
    "    j = round(len(L)/nchunks)\n",
    "    chunk_size = j\n",
    "    i = 0\n",
    "    while i < len(L):\n",
    "        chunk = L[i:j]\n",
    "        L2.append(chunk)\n",
    "        i = j\n",
    "        j += chunk_size\n",
    "        if(j>len(L)):\n",
    "            j = len(L)\n",
    "    return L2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def worker(input_chunk):\n",
    "    ref_alt_library = VCF(myinputvcf)\n",
    "    \n",
    "    lines = []\n",
    "    \n",
    "    for chrom_pos in input_chunk:\n",
    "        chromie = chrom_pos[0]\n",
    "        base_position = chrom_pos[1]\n",
    "        target_row = data[(data['chromosome'] == chromie) & (data['position'] == base_position)]\n",
    "        \n",
    "        allele1 = target_row['allele1']\n",
    "        allele1 = str(allele1.values)\n",
    "        allele1 = allele1.replace(\"['\",\"\")\n",
    "        allele1 = allele1.replace(\"']\",\"\")\n",
    "        \n",
    "        allele2 = target_row['allele2\\r']\n",
    "        allele2 = str(allele2.values)\n",
    "        allele2 = allele2.replace(\"['\",\"\")\n",
    "        allele2 = allele2.replace(\"']\",\"\")\n",
    "        \n",
    "        position = str(chromie) + \":\" + str(base_position) + \"-\" + str(base_position)\n",
    "        \n",
    "        for record in ref_alt_library(position):\n",
    "            read = str(record).split()\n",
    "            chrom, bp, rsid, REF, ALT = read[0], read[1], read[2], read[3], read[4]\n",
    "            \n",
    "            \n",
    "            if str(allele1) == str(REF) and str(allele2) == str(REF):\n",
    "                genotypes = '0/0'\n",
    "            elif str(allele1) == str(REF) and str(allele2) == str(ALT):\n",
    "                genotypes = '0/1'\n",
    "            elif str(allele1) == str(ALT) and str(allele2) == str(REF):\n",
    "                genotypes = '1/0'\n",
    "            elif str(allele1) == str(ALT) and str(allele2) == str(ALT):\n",
    "                genotypes = '1/1'\n",
    "            else:\n",
    "                genotypes = '.'\n",
    "                \n",
    "            line =[chrom, bp, rsid, REF, ALT, '.', '.', '.', 'GT', genotypes]\n",
    "            \n",
    "            lines.append(line)\n",
    "            \n",
    "    return lines   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_get_ref_alt(df):\n",
    "    \n",
    "    results = []\n",
    "    \n",
    "    data_2 = add_columns(df)\n",
    "\n",
    "    for currChrom in range(1,23):\n",
    "        print('We are starting chromosome' + str(currChrom))\n",
    "        global myinputvcf\n",
    "        myinputvcf = '/gpfs/group/torkamani/shaun/mygenerank-pipeline_beta/conversion/1000G_HG00096/ALL.chr'+str(currChrom)+'.phase3.HG00096.vcf.gz'\n",
    "\n",
    "        coll = data[data['chromosome'] == currChrom]\n",
    "        target_coll = coll.loc[:, 'chromosome' : 'position']\n",
    "        target_list = target_coll.values.tolist()\n",
    "\n",
    "        chunks = chunk(target_list, ncores)\n",
    "        pool = multiprocessing.Pool(ncores)\n",
    "        temp_results = pool.map(worker, chunks)\n",
    "        \n",
    "        pool.close()\n",
    "        pool.join()\n",
    "\n",
    "        temp_results = [item for sublist in temp_results for item in sublist]\n",
    "        results += temp_results\n",
    "        \n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "began\n",
      "We are starting chromosome1\n",
      "We are starting chromosome2\n",
      "We are starting chromosome3\n",
      "We are starting chromosome4\n",
      "We are starting chromosome5\n",
      "We are starting chromosome6\n",
      "We are starting chromosome7\n",
      "We are starting chromosome8\n",
      "We are starting chromosome9\n",
      "We are starting chromosome10\n",
      "We are starting chromosome11\n",
      "We are starting chromosome12\n",
      "We are starting chromosome13\n",
      "We are starting chromosome14\n",
      "We are starting chromosome15\n",
      "We are starting chromosome16\n",
      "We are starting chromosome17\n",
      "We are starting chromosome18\n",
      "We are starting chromosome19\n",
      "We are starting chromosome20\n",
      "We are starting chromosome21\n",
      "We are starting chromosome22\n",
      "Conversion Time:\n",
      "249.01769852638245\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "print('began')\n",
    "\n",
    "myinputtxt = 'AncestryDNA2.txt'\n",
    "original_file = pd.read_csv(myinputtxt, sep = '\\t', lineterminator='\\n', comment='#')\n",
    "data = (original_file[original_file[\"chromosome\"] <= 22])\n",
    "\n",
    "ncores = multiprocessing.cpu_count()\n",
    "\n",
    "\n",
    "sub_dict = {}\n",
    "myinputvcf = \"\"\n",
    "\n",
    "outputfilename = myinputtxt.strip('.txt') + '.vcf'\n",
    "with open(outputfilename, 'w') as outputfile:\n",
    "\n",
    "    today = date.today().strftime('%Y%m%d')\n",
    "    header = ['##fileformat=VCFv4.2',\n",
    "              '##fileDate=' + today,\n",
    "              '##source=conversion_from_Ancestry_user_download',\n",
    "              '##reference=1000Genomes',\n",
    "              '##INFO=<ID=ID,Number=1,Type=Integer,Description=\"N/A\">',\n",
    "              '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"N/A\">',\n",
    "              '##FORMAT=<ID=NA,Number=1,Type=String,Description=\"N/A\">']\n",
    "    header = \"\\n\".join(header)\n",
    "\n",
    "\n",
    "\n",
    "    finaldata = test_get_ref_alt(data)\n",
    "    end_data = pd.DataFrame(finaldata, columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'GENOTYPE'])\n",
    "\n",
    "    outputfile.write(header + '\\n')\n",
    "    outputfile.write(end_data.to_string())\n",
    "\n",
    "\n",
    "end = time.time()\n",
    "\n",
    "print('Conversion Time:')\n",
    "print(end - start)\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
